src/daft-recordbatch/src/ops/window_states/sum.rs

```diff
@@ -0,0 +1,132 @@
+use std::{
+    marker::PhantomData,
+    ops::{Add, Sub},
+};
+
+use arrow2::bitmap::MutableBitmap;
+use common_error::{DaftError, DaftResult};
+use daft_core::{datatypes::DaftPrimitiveType, prelude::*};
+use num_traits::Zero;
+
+use super::WindowAggStateOps;
+
+pub struct SumWindowState<'a, T>
+where
+    T: DaftPrimitiveType,
+    T::Native: Zero + Add<Output = T::Native> + Sub<Output = T::Native> + Copy,
+{
+    source: DataArray<T>,
+    sum: T::Native,
+    sum_vec: Vec<T::Native>,
+    valid_count: usize,
+    validity: MutableBitmap,
+    _phantom: PhantomData<&'a ()>,
+}
+
+impl<'a, T> SumWindowState<'a, T>
+where
+    T: DaftPrimitiveType,
+    T::Native: Zero + Add<Output = T::Native> + Sub<Output = T::Native> + Copy,
+{
+    pub fn new(source: &'a Series, total_length: usize) -> DaftResult<Self> {
+        let source_array = source.downcast::<DataArray<T>>().cloned().or_else(
+            |_| -> Result<DataArray<T>, DaftError> {
+                let target_type = T::get_dtype();
+                source
+                    .cast(&target_type)
+                    .and_then(|casted| casted.downcast::<DataArray<T>>().cloned())
+            },
+        )?; // unsure whether or_else would ever trigger
+
+        Ok(Self {
+            source: source_array,
+            sum: T::Native::zero(),
+            sum_vec: Vec::with_capacity(total_length),
+            valid_count: 0,
+            validity: MutableBitmap::with_capacity(total_length),
+            _phantom: PhantomData,
+        })
+    }
+}
+
+impl<'a, T> WindowAggStateOps<'a> for SumWindowState<'a, T>
+where
+    T: DaftPrimitiveType,
+    T::Native: Zero + Add<Output = T::Native> + Sub<Output = T::Native> + Copy,
+    DataArray<T>: IntoSeries,
+{
+    fn add(&mut self, start_idx: usize, end_idx: usize) -> DaftResult<()> {
+        assert!(
+            end_idx > start_idx,
+            "end_idx must be greater than start_idx"
+        );
+
+        for i in start_idx..end_idx {
+            if self.source.is_valid(i) {
+                self.sum = self.sum + self.source.get(i).unwrap();
```

Comment:
I suspect that this will be an issue for NaNs, since when you encounter a NaN, adding it to `self.sum` will result in NaN, which is fine, but when you try to remove it, `NaN - NaN = NaN`, so it's going to be NaNs for the rest of the partition

Reply:
Let's also add a test for this